### PR TITLE
[contracts] make `debug_message` execution outcome invariant to node debug logging setting

### PR DESCRIPTION
**Pull request by agryaznov**
Follow-up to https://github.com/paritytech/substrate/pull/12953

FIFO was not really a good approach to the debug buffer, as it's not very efficient in computational sense. This PR targets to improve it as follows: 

- the function just silently drops the message if no space left in the debug buffer, or when the message is not a valid utf-8 string; 
- basically we make it indistinguishable from contract side, whether the message got logged or not;
- the gas for the function call is to be charged per byte of the passed message; updated benchmarks;

### CODE REVIEW

File: frame/contracts/src/tests.rs
Comment (athei) [inline frame/contracts/src/tests.rs:2717]: We should add an additional check here that the buffer is empty.

File: frame/contracts/src/wasm/mod.rs
Comment (athei) [inline frame/contracts/src/wasm/mod.rs:2297]: Let's check that the buffer is empty.

File: frame/contracts/src/benchmarking/mod.rs
Comment (athei) [inline frame/contracts/src/benchmarking/mod.rs:925]: You should only let `i` run up until the amount of memory you are able to initialise. Otherwise the vast amount of sizes are against invalid sequence `0`. This will not copy it to the buffer and takes less time. Hence we underestimate. In benchmarks we should always benchmark the worst case. Hence only initialise with valid characters which will be copied to the buffer and incur the biggest cost.

Just let `i` run up to the allowed buffer size capped by the max code size. We probably can't use batching here and just do one run (otherwise the buffer might fill up).

File: frame/contracts/src/wasm/runtime.rs
Comment (athei) [inline frame/contracts/src/wasm/runtime.rs:2369]: Let's cap the `str_len` to the maximum buffer bound as the first thing in this function.

File: frame/contracts/src/benchmarking/mod.rs
Comment (agryaznov) [reply frame/contracts/src/benchmarking/mod.rs]: > Are you sure? I think benchmarks skip the size and other validation steps.

Yep, it [fails with `Original code too large`](https://github.com/paritytech/substrate/blob/ag-debug-buffer/frame/contracts/src/wasm/prepare.rs#L566) when preparing the wasm module (only when it's really too large ofc)

> You should only let `i` run up until the amount of memory you are able to initialise. Otherwise the vast amount of sizes are against invalid sequence `0`. This will not copy it to the buffer and takes less time. Hence we underestimate. In benchmarks we should always benchmark the worst case. Hence only initialise with valid characters which will be copied to the buffer and incur the biggest cost.
> 
> We probably can't use batching here because the buffer will be full in the later runs.


Agreed. Updated the benchmark.

**Comment by agryaznov** [comment]
/cmd queue -c bench-bot $ pallet dev pallet_contracts

**Comment by agryaznov** [comment]
/cmd queue -c bench $ pallet dev pallet_contracts

File: frame/contracts/src/benchmarking/mod.rs
Comment (athei) [inline frame/contracts/src/benchmarking/mod.rs:918]: This should only run up to the amount you actually initialised: `T::MaxCodeLen::get() / 1024 - 1`

File: frame/contracts/src/benchmarking/mod.rs
Comment (athei) [inline frame/contracts/src/benchmarking/mod.rs:923]:
```suggestion
		// We benchmark versus messages containing printable ASCII codes.
		// About 1Kb goes to the instrumented contract code instructions,
		// whereas all the space left we use for the initialization of the debug messages data.
		let message = (0 .. T::MaxCodeLen::get() - 1024).zip(32..127).cycle()).map(|i| i.1).collect::<Vec<_>>();
```

File: frame/contracts/src/benchmarking/mod.rs
Comment (athei) [inline frame/contracts/src/benchmarking/mod.rs:938]:
```suggestion
					value: message,
```

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2350452

File: frame/contracts/src/benchmarking/mod.rs
Comment (agryaznov) [reply frame/contracts/src/benchmarking/mod.rs:918]: Why? All uninitialized bytes get `0` value, which is a proper UTF-8 code (`NUL`), so we can actually test versus longer inputs having it the way it is now. 

We just have to initialize a smaller part because of the contracts code length restriction:  the initializers (data segments) for the whole contract memory would just make the benchmark-generated contract size exceed the `MaxCodeLen` limit. That's being said, the debug message input could really exceed the size we initialize here in the benchmark.

Comment (athei) [reply frame/contracts/src/benchmarking/mod.rs:952]: You are right. I was just afraid that it stops parsing at the first zero byte because of null termination. But Rust ist UTF-8 compliant and should allow nulls within strings.

File: frame/contracts/src/benchmarking/mod.rs
Comment (athei) [inline frame/contracts/src/benchmarking/mod.rs:974]: You need to use `body::repeated(API_BENCHMARK_BATCH_SIZE, ...` here since the schedule divides it by the batch size.

File: frame/contracts/src/schedule.rs
Comment (athei) [inline frame/contracts/src/schedule.rs:650]: We don't batch in the benchmark.

```suggestion
			debug_message_per_byte: to_weight!(cost_byte!(seal_debug_message_per_kb)),
```

File: frame/contracts/src/benchmarking/mod.rs
Comment (agryaznov) [reply frame/contracts/src/benchmarking/mod.rs:952]: True, the host function just [cuts the passed message](https://github.com/paritytech/substrate/blob/ag-debug-buffer/frame/contracts/src/wasm/runtime.rs#L2385) to the `MaxDebugBufferLen` size. So if we had that bound less than maximum allowed contract memory (which is not the case currently), the benchmark result would be distorted. I corrected the benchmark to vary its input up to whichever is less among of the two bounds.

File: frame/contracts/src/schedule.rs
Comment (agryaznov) [reply frame/contracts/src/schedule.rs:650]: True. Fixed.

**Review by athei — APPROVED** [review]
We need to re-run the benchmarks with the proper machine before we merge, though.

**Comment by agryaznov** [comment]
looks like we're waiting new weights from here: https://github.com/paritytech/substrate/pull/13316

**Comment by koute** [comment]
The extra commit I've just pushed is a workaround for this issue: https://github.com/paritytech/substrate/issues/13376

It replaces the dependency on `wasmtime` with our own fork with the fix applied (which is trivial). This is only temporary until (hopefully) it's fixed upstream in `wasmtime`. (But it should be fine to merge it as-is, assuming the CI will be happy with it.)

**Comment by alvicsam** [comment]
bot bench -v PIPELINE_SCRIPTS_REF=bm-fallback $ pallet dev pallet_contracts

**Comment by athei** [comment]
bot merge

**Pull request merged**
